clang/tools/clang-linker-wrapper/SYCLOffloadWrapper.cpp

```diff
@@ -0,0 +1,748 @@
+  }
+
+  // Emits an object that holds <address, size> pair for the device image
+  // and puts it into a .tgtimg section. This section can be used for
+  // finding and extracting all device images from the fat binary after
+  // linking.
+  void emitRegistrationFunctions(Constant *Address, size_t Size, size_t ImageID,
+                                 StringRef OffloadKindTag) {
+    Type *IntPtrTy = M.getDataLayout().getIntPtrType(C);
+    auto *ImgInfoArr =
+        ConstantArray::get(ArrayType::get(IntPtrTy, 2),
+                           {ConstantExpr::getPointerCast(Address, IntPtrTy),
+                            ConstantInt::get(IntPtrTy, Size)});
+    auto *ImgInfoVar = new GlobalVariable(
+        M, ImgInfoArr->getType(), true, GlobalVariable::InternalLinkage,
+        ImgInfoArr, Twine(OffloadKindTag) + Twine(ImageID) + Twine(".info"));
+    ImgInfoVar->setAlignment(
+        MaybeAlign(M.getDataLayout().getTypeStoreSize(IntPtrTy) * 2u));
+    ImgInfoVar->setUnnamedAddr(GlobalValue::UnnamedAddr::Local);
+    ImgInfoVar->setSection(".tgtimg");
+
+    // Add image info to the used list to force it to be emitted to the
+    // object.
+    appendToUsed(M, ImgInfoVar);
+  }
+
+  Constant *wrapImage(const SYCLImage &Image, size_t ImageID,
+                      StringRef OffloadKindTag) {
+    auto *NullPtr = Constant::getNullValue(PointerType::getUnqual(C));
+    // DeviceImageStructVersion change log:
+    // -- version 2: updated to PI 1.2 binary image format
+    constexpr uint16_t DeviceImageStructVersion = 2;
+    constexpr uint8_t SYCLOffloadKind = 4;
+    auto *Version =
+        ConstantInt::get(Type::getInt16Ty(C), DeviceImageStructVersion);
+    auto *Kind = ConstantInt::get(Type::getInt8Ty(C), SYCLOffloadKind);
+    auto *Format = ConstantInt::get(
+        Type::getInt8Ty(C),
+        static_cast<int8_t>(binaryImageFormatToInt8(Image.Format)));
+    auto *Target =
+        addStringToModule(Image.Target, Twine(OffloadKindTag) +
+                                            Twine("target.") + Twine(ImageID));
+    auto *CompileOptions = addStringToModule(
+        Image.CompileOptions,
+        Twine(OffloadKindTag) + Twine("opts.compile.") + Twine(ImageID));
+    auto *LinkOptions = addStringToModule(
+        Image.LinkOptions,
+        Twine(OffloadKindTag) + Twine("opts.link.") + Twine(ImageID));
+
+    std::pair<Constant *, Constant *> PropSets =
+        addPropertySetRegistry(Image.PropertyRegistry);
+
+    std::pair<Constant *, Constant *> Binary;
+    if (Image.Target == "native_cpu")
+      Binary = addDeclarationsForNativeCPU(Image.Entries);
+    else {
+      Binary = addDeviceImageToModule(
+          Image.Image, Twine(OffloadKindTag) + Twine(ImageID) + Twine(".data"),
+          Image.Target);
+    }
+
+    // Note: Manifests are deprecated but corresponding nullptr fields should
+    // remain to comply ABI.
+    std::pair<Constant *, Constant *> Manifests = {NullPtr, NullPtr};
```

Comment:
We should use this switch to the new offloading model as an opportunity to cleanup all legacy we have. As long as RT supports both old and new formats of device images, we are free to just always generate a new format here.